### PR TITLE
feat(QueryBuilder): add method clear(statement)

### PR DESCRIPTION
**Pull request by intech**
Add method `clear(statement)` in `QueryBuilder` for https://github.com/knex/knex/issues/1476

#### TODO:
- [x] approve with the community
- [x] docs
- [x] tests

### CODE REVIEW

**Comment by kibertoad** [comment]
Thank you for the contribution! Could you please add tests for this?

**Comment by intech** [comment]
> Thank you for the contribution! Could you please add tests for this?

Oh sure!
I am worried that I just got acquainted with the project three days ago and really need your approval. Did I take everything into account in logic?

File: lib/query/builder.js
Comment (kibertoad) [inline lib/query/builder.js:1035]: Why mark existing methods as deprecated?

File: lib/query/builder.js
Comment (kibertoad) [inline lib/query/builder.js:1045]: Could you please add `{}`?

File: lib/query/builder.js
Comment (kibertoad) [inline lib/query/builder.js:1254]: Is there a particular reason to use `delete` over "set to undefined" here? Delete is slow.

File: lib/query/builder.js
Comment (intech) [reply lib/query/builder.js:1035]: For further refactoring, remove duplicate features.
Leave backward compatibility to major version.
What do you think?

File: lib/query/builder.js
Comment (intech) [reply lib/query/builder.js:1045]: Specify what it is about?

Comment (kibertoad) [reply lib/query/builder.js:1045]: `if (xyz) {  statement }`

**Comment by intech** [comment]
@kibertoad Fixed the code. Added a test. Updated the documentation.
Please review. Thanks!

**Comment by intech** [comment]
@kibertoad I propose to merge so that people can already use.
And this PR for docs https://github.com/knex/documentation/pull/289

**Comment by intech** [comment]
@kibertoad we are still waiting for your approval :hourglass_flowing_sand:

File: lib/query/builder.js
Comment (kibertoad) [reply lib/query/builder.js:1035]: Isn't there a difference between those methods, though? From my understanding, existing ones are more granular, while new ones clears everything. Do I read code wrong?

Comment (intech) [reply lib/query/builder.js:1035]: `clear("order")` method completely replaces `clearOrder()`, the difference is that clear gives you more options for limit, offset and other

File: test/unit/query/builder.js
Comment (elhigu) [inline test/unit/query/builder.js:746]: More tests needed to make sure that each clear statement works separately and doesn't clear anything unexpected + test for non-happy-path while using non existing statement.  

Also some test showing that it does not clear anything from subquery would be good to have.

Comment (intech) [reply test/unit/query/builder.js:746]: I agree, I'll do it.

File: lib/query/builder.js
Comment (kibertoad) [inline lib/query/builder.js:1011]: Can you please add non-deprecated alternative in comments to every deprecated method?

Comment (kibertoad) [reply lib/query/builder.js:1011]: @elhigu Do you think we should output deprecation warning here as well?

Comment (elhigu) [reply lib/query/builder.js:1011]: I think those separate clear methods can be deprecated as long as list of valid string literals are listed in typings.

**Review by kibertoad — CHANGES_REQUESTED** [review]
Could you also update types in https://github.com/knex/knex/blob/master/types/index.d.ts?

**Comment by kibertoad** [comment]
@intech Could you please update Typescript types? then it should be good to be merged.

**Comment by intech** [comment]
@kibertoad I'll do it along with additional tests. A little more patience. Thanks.

File: types/index.d.ts
Comment (kibertoad) [inline types/index.d.ts:536]: Could you narrow down the type of `statement` to the list of supported statements?

Comment (intech) [reply types/index.d.ts]: @kibertoad I don't know typescript, I hope I did it right. I would be grateful for your review.

File: types/index.d.ts
Comment (elhigu) [inline types/index.d.ts:427]: Naming this ClearStatements would be less probably to conflict with some other random type declarations.

Comment (elhigu) [reply types/index.d.ts:427]: Other that that this looks what I was expecting 👍

**Review by elhigu — APPROVED** [review]
Looks good to me. Was the documentation up to date?

**Comment by intech** [comment]
> Looks good to me. Was the documentation up to date?

@elhigu Check https://github.com/knex/documentation/pull/289/files#diff-ca136b3642dd50189c1489da6c71328577609b9df883bbe284e739c2de7a7660R1110

**Comment by elhigu** [comment]
> > Looks good to me. Was the documentation up to date?
> 
> @elhigu Check https://github.com/knex/documentation/pull/289/files#diff-ca136b3642dd50189c1489da6c71328577609b9df883bbe284e739c2de7a7660R1110

Old methods could be marked as deprecated 🤔

**Comment by intech** [comment]
@elhigu there is not a single deprecated method in the documentation, I do not know how best to do it. Do you have ideas?

**Comment by intech** [comment]
Perhaps this option will work?

**Comment by elhigu** [comment]
Looks perfect 👍

**Comment by intech** [comment]
@elhigu in the commit code the menu item is added. In the screenshot I made a visualization in devtools to see :)

**Comment by intech** [comment]
It will remain for me to cover all statements with tests, I doubt that `clearCounter()` will work correctly there
https://github.com/knex/knex/blob/993c2104c54a49dddf3cee8e1dc59a2d54282972/lib/query/builder.js#L969-L970

is cleared and
https://github.com/knex/knex/blob/993c2104c54a49dddf3cee8e1dc59a2d54282972/lib/query/builder.js#L1253-L1255

occurs in the `_clearGroups ()` method.
I think it will throw an error. I recommend waiting for the test coverage. I'll do it by the end of the week.

**Comment by intech** [comment]
I think everything.
There is a separate situation with the `.clearCounters()` method I removed the description of `deprecated` on it.
In the new method `.clear()` added 2 more values ​​for the parameter `counter` and `counters`, both are aliases for the method `.clearCounters()`.

It looks right in my opinion, I didn't complicate one method, but added sugar to it.

**Review by kibertoad — APPROVED** [review]

**Comment by kibertoad** [comment]
Thanks a lot!

**Comment by kibertoad** [comment]
Released in 0.21.9

**Pull request merged**
